pinot-core/src/main/java/org/apache/pinot/core/query/executor/TableExecutionInfo.java

```diff
@@ -0,0 +1,211 @@
+
+  /**
+   * Get the number of segments acquired.
+   * @return The number of segments acquired
+   */
+  int getNumSegmentsAcquired();
+
+  /**
+   * Get the selected segments and segment contexts for a table referenced in a query. The information is gathered
+   * in a SelectSegmentsInfo object.
+   * @return A SelectSegmentsInfo object containing the selected segments and segment contexts
+   */
+  SelectedSegmentsInfo getSelectedSegmentsInfo(QueryContext queryContext, TimerContext timerContext,
+      ExecutorService executorService, SegmentPrunerService segmentPrunerService);
+
+  /**
+   * Generate metadata about the consuming segments queried.
+   * @return A ConsumingSegmentsInfo object containing the metadata about the consuming segments queried
+   */
+  ConsumingSegmentsInfo getConsumingSegmentsInfo();
+
+  /**
+   * If consuming segments are being queried, this class contains the information about the consuming segments such as
+   * the number of segments queried, the min index time, the min ingestion time and the max end time.
+   */
+  class ConsumingSegmentsInfo {
+    private final int _numConsumingSegmentsQueried;
+    private final long _minIndexTimeMs;
+    private final long _minIngestionTimeMs;
+    private final long _maxEndTimeMs;
+
+    public ConsumingSegmentsInfo(int numConsumingSegmentsQueried, long minIndexTimeMs, long minIngestionTimeMs,
+        long maxEndTimeMs) {
+      _numConsumingSegmentsQueried = numConsumingSegmentsQueried;
+      _minIndexTimeMs = minIndexTimeMs;
+      _minIngestionTimeMs = minIngestionTimeMs;
+      _maxEndTimeMs = maxEndTimeMs;
+    }
+
+    long getMinConsumingFreshnessTimeMs() {
+      long minConsumingFreshnessTimeMs = 0;
+      if (getMinIngestionTimeMs() != Long.MAX_VALUE) {
+        minConsumingFreshnessTimeMs = getMinIngestionTimeMs();
+      } else if (getMinIndexTimeMs() != Long.MAX_VALUE) {
+        minConsumingFreshnessTimeMs = getMinIndexTimeMs();
+      } else if (getMaxEndTimeMs() != Long.MIN_VALUE) {
+        minConsumingFreshnessTimeMs = getMaxEndTimeMs();
+      }
+      return minConsumingFreshnessTimeMs;
+    }
+
+    public int getNumConsumingSegmentsQueried() {
+      return _numConsumingSegmentsQueried;
+    }
+
+    public long getMinIndexTimeMs() {
+      return _minIndexTimeMs;
+    }
+
+    public long getMinIngestionTimeMs() {
+      return _minIngestionTimeMs;
+    }
+
+    public long getMaxEndTimeMs() {
+      return _maxEndTimeMs;
+    }
+  }
+
+  /**
+   * This class contains the information about the selected segments such as the number of segments queried, the
+   * number of segments selected and the number of total documents.
+   */
+  class SelectedSegmentsInfo {
+    private List<IndexSegment> _indexSegments;
+    private long _numTotalDocs;
+    private SegmentPrunerStatistics _prunerStats;
+    private int _numTotalSegments;
+    private int _numSelectedSegments;
+    private List<SegmentContext> _selectedSegmentContexts;
```

Comment:
nit: can be `final`?

Reply:
No. For logical tables, I will add this function to aggregate information across many tables.

```
 public void aggregate(SelectedSegmentsInfo other) {
      _indexSegments.addAll(other._indexSegments);
      _numTotalDocs += other._numTotalDocs;
      _numTotalSegments += other._numTotalSegments;
      _numSelectedSegments += other._numSelectedSegments;
      _selectedSegmentContexts.addAll(other._selectedSegmentContexts);
      _prunerStats.setInvalidSegments(_prunerStats.getInvalidSegments() + other._prunerStats.getInvalidSegments());
      _prunerStats.setValuePruned(_prunerStats.getValuePruned() + other._prunerStats.getValuePruned());
      _prunerStats.setLimitPruned(_prunerStats.getLimitPruned() + other._prunerStats.getLimitPruned());
    }
```